File: frontend/src/types/aiReview.ts
/**
 * R20: AI 預審與驗證相關型別
 */

// Level 1 驗證結果
export interface ValidationIssue {
    code: string
    category: string
    section: string
    message: string
    suggestion: string
}

export interface ValidationResult {
    passed: string[]
    errors: ValidationIssue[]
    warnings: ValidationIssue[]
}

// AI 預審結果
export interface AiReviewIssue {
    severity: 'error' | 'warning'
    category: string
    section: string
    message: string
    suggestion: string
}

export interface AiReviewAiResult {
    summary: string
    score?: number
    issues: AiReviewIssue[]
    passed: string[]
}

// 執行秘書標註
export interface StaffReviewFlag {
    flag_type: 'needs_attention' | 'concern' | 'suggestion'
    section: string
    message: string
    suggestion: string
}

export interface StaffAiResult {
    summary: string
    flags: StaffReviewFlag[]
}

// API Response
export interface AiReviewResponse {
    id: string
    protocol_id: string
    review_type: 'client_pre_submit' | 'staff_pre_review'
    rule_result?: ValidationResult
    ai_result?: AiReviewAiResult | StaffAiResult
    ai_model?: string
    total_errors: number
    total_warnings: number
    score?: number
    duration_ms?: number
    created_at: string
}

export interface RemainingCount {
    remaining: number
}

// 批次退回補件
export interface BatchReturnFlag {
    section: string
    message: string
    suggestion: string
}

export interface BatchReturnRequest {
    flags: BatchReturnFlag[]
    additional_note?: string
}

export interface BatchReturnResponse {
    created_comments: number
    status: string
}

File: backend/src/models/ai_review.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use utoipa::ToSchema;
use uuid::Uuid;

// ── DB Entity ──

#[derive(Debug, Clone, FromRow)]
pub struct ProtocolAiReview {
    pub id: Uuid,
    pub protocol_id: Uuid,
    pub protocol_version_id: Option<Uuid>,
    pub review_type: String,
    pub rule_result: Option<serde_json::Value>,
    pub ai_result: Option<serde_json::Value>,
    pub ai_model: Option<String>,
    pub ai_input_tokens: Option<i32>,
    pub ai_output_tokens: Option<i32>,
    pub total_errors: i32,
    pub total_warnings: i32,
    pub score: Option<i32>,
    pub triggered_by: Option<Uuid>,
    pub duration_ms: Option<i32>,
    pub created_at: DateTime<Utc>,
}

// ── Validation Result (Level 1) ──

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct ValidationResult {
    pub passed: Vec<String>,
    pub errors: Vec<ValidationIssue>,
    pub warnings: Vec<ValidationIssue>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct ValidationIssue {
    pub code: String,
    pub category: String,
    pub section: String,
    pub message: String,
    pub suggestion: String,
}

// ── AI Review Result (Level 2) ──

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct AiReviewIssue {
    pub severity: String,
    pub category: String,
    pub section: String,
    pub message: String,
    pub suggestion: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct AiReviewAiResult {
    pub summary: String,
    pub score: Option<i32>,
    pub issues: Vec<AiReviewIssue>,
    pub passed: Vec<String>,
}

/// 執行秘書標註 flag
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct StaffReviewFlag {
    pub flag_type: String,
    pub section: String,
    pub message: String,
    pub suggestion: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct StaffAiResult {
    pub summary: String,
    pub flags: Vec<StaffReviewFlag>,
}

// ── Batch Return Request ──

#[derive(Debug, Deserialize, ToSchema)]
pub struct BatchReturnFlag {
    pub flag_type: String,
    pub section: String,
    pub message: String,
    pub suggestion: String,
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct BatchReturnRequest {
    pub ai_review_id: Uuid,
    pub flags: Vec<BatchReturnFlag>,
    pub additional_note: Option<String>,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct BatchReturnResponse {
    pub created_comments: usize,
    pub status: String,
}

// ── API Response ──

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct AiReviewResponse {
    pub id: Uuid,
    pub protocol_id: Uuid,
    pub review_type: String,
    pub rule_result: Option<ValidationResult>,
    pub ai_result: Option<serde_json::Value>,
    pub ai_model: Option<String>,
    pub total_errors: i32,
    pub total_warnings: i32,
    pub score: Option<i32>,
    pub duration_ms: Option<i32>,
    pub created_at: DateTime<Utc>,
}

impl From<ProtocolAiReview> for AiReviewResponse {
    fn from(r: ProtocolAiReview) -> Self {
        let rule_result = r.rule_result.and_then(|v| serde_json::from_value(v).ok());
        Self {
            id: r.id,
            protocol_id: r.protocol_id,
            review_type: r.review_type,
            rule_result,
            ai_result: r.ai_result,
            ai_model: r.ai_model,
            total_errors: r.total_errors,
            total_warnings: r.total_warnings,
            score: r.score,
            duration_ms: r.duration_ms,
            created_at: r.created_at,
        }
    }
}
